### PR TITLE
Bookmarks sync evolution

### PR DESCRIPTION
**Pull request by bendk**
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/places/src/bookmark_sync/tests/payload_evolution.rs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('c'), 2, vec![("device_type", "desktop")]),
+            // We re-upload query and livemark records after a reset, so test them here
+            // ... except for some reason apply_incoming() flags these as deleted and wants to sync
+            // outgoing tombstones for them
+            // remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
+            // remote_livemark(guid('e'), vec![("device_type", "desktop")]),
```
Comment (bendk) [inline components/places/src/bookmark_sync/tests/payload_evolution.rs:100]: I really wanted to test this case, but I couldn't figure out this issue.  I'm hoping @mhammond knows what's going on here.

Comment (linabutler) [reply components/places/src/bookmark_sync/tests/payload_evolution.rs:100]: You actually uncovered an intentional behavior! Livemarks were [removed in Firefox Desktop 64](https://support.mozilla.org/en-US/kb/feed-reader-replacements-firefox), and none of our other platforms ever supported them, so Sync deletes them.

There's a longer saga behind livemarks if you're curious! 😄 Forever ago, Desktop's Places stored them as folders with special annotations—the feed and site URLs—and real bookmark children for each feed item. Sync had a special record type for them, too, but otherwise treated them as regular folders. That scheme caused write and sync churn, though, so livemarks became empty folders with annotations, and the items were populated dynamically in the UI.

Some of our power users loved livemarks, but they never became a super popular feature, and couldn't justify their maintenance burden—so the Desktop team decided to remove them in 64. That release had a migration step that exported livemarks to OPML, then deleted them from Places.

We didn't want Sync to resurrect them—or turn them into folders—so we also added logic in [the merger](https://github.com/mozilla/dogear/pull/37) and [Desktop's mirror](https://bugzilla.mozilla.org/show_bug.cgi?id=1546035) to delete them.

TL;DR: Rust Places inherited the logic to delete livemarks from Desktop. Even though Fenix and iOS never supported livemarks, they'll still upload tombstones if they ever see one.

Since they're not a thing anymore, I think it's also totally OK not to support unknown fields for them—it doesn't hurt to, but it is extra code that none of our platforms should ever invoke.

I hope that helps!

Comment (bendk) [reply components/places/src/bookmark_sync/tests/payload_evolution.rs:100]: That's definitely helpful, thanks for the context here.

Do you know if queries are in the same category or are they still supported?

File: components/places/src/bookmark_sync/tests/payload_evolution.rs [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+                guid('b'),
+                "Animals",
+                vec!["A", "B"],
+                vec![("device_type", "phone"), ("color", "blue")],
+            ),
+            remote_separator(guid('c'), 2, vec![("device_type", "desktop")]),
+            // We re-upload query and livemark records after a reset, so test them here
+            // ... except for some reason apply_incoming() flags these as deleted and wants to sync
+            // outgoing tombstones for them
+            // remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
```
Comment (linabutler) [inline components/places/src/bookmark_sync/tests/payload_evolution.rs:99]: Here's the test fix for queries (plus plumbing `store_incoming_query` / `BookmarkItemRecord::unknown_fields`):

```suggestion
            remote_item(
                json!({
                    "id": "unfiled",
                    "title": "Unfiled Bookmarks",
                    "type": "folder",
                    "parentid": "places",
                    "parentName": "places",
                    "dateAdded": 0,
                    "children": vec![guid('d')],
                }),
                vec![],
            ),
            remote_query(guid('d'), "dog search", "place:folder=123&excludeItems=1", vec![("device_type", "desktop")]),
```

I'll post the explanation in a bit!

Comment (linabutler) [reply components/places/src/bookmark_sync/tests/payload_evolution.rs:99]: Queries are still supported in Desktop Places. Fenix and iOS don't show them in the UI, but Rust Places still stores and round-trips them. There are three special cases in Sync's handling of queries:

1. Rewriting tag folder queries to reference tag names instead of folders.
2. Rewriting non-tag `folder` queries to add `excludeItems`.
3. Removing suspected "orphaned left pane queries".

I think your test unintentionally triggered (3), which is why it was uploading a tombstone.

What's a left pane query? If you open Bookmarks > Manage Bookmarks in Firefox Desktop, you'll see a two-pane window, with "All Bookmarks", "Bookmarks Toolbar", "Bookmarks Menu", "Other Bookmarks", and "Mobile Bookmarks" in the left pane. These left-pane queries show the contents of those roots in Places. Since 2017-ish, these queries are implemented entirely in the UI.

Before 2017, these left-pane queries were stored in Places as real query items, as children and grandchildren of the "left pane root". This root was hidden from the UI, and not synced. Before our Desktop bookmarks tracker and mirror rewrites, though, it was possible in some cases for Sync to observe changes to those left pane queries–like localized titles, or query URLs, changing between releases—and upload records for them. Crucially, it never uploaded the left pane root, so receiving devices would see those left pane queries, decide they were orphaned, and move them into `unfiled`.

This caused all sorts of issues in the UI. In the worst case, expanding one of those orphaned left pane queries at the wrong time would cause an infinite loop. We didn't have a good way of identifying these queries—their URLs and titles could change between releases, and the left pane root had a random GUID per device—so we used a [heuristic](https://github.com/mozilla/dogear/blob/f83e6db9640332bb4d8b9689740a7db68ba087f3/src/tree.rs#L1583-L1584). **If we saw any orphaned query—where it had a `parentid` that either didn't exist, or didn't mention the query in its `children` array—we'd delete it.**

TL;DR: Your test was accidentally creating one of these orphaned queries: `remote_query()` returns a record with `"parentid: "unfiled"`, but `initial_remote_records` didn't have a record for `unfiled`, so Dogear assumed it was an orphaned left pane query, and deleted it.

Your test didn't trigger the other two special cases, but here's the explanation for completeness.

(1) and (2) have the same underlying cause: the `folder` param references the SQLite row ID, which isn't portable across devices. When a query references a nonexistent `folder` ID, it'll return _all_ bookmarks in the database, which breaks the UI in the same way as (3).

To fix (1), Sync rewrites tag queries (with `type=7`) from `folder=<tag folder ID>` to `tag=<tag name>`. Desktop Places added support for `tag=` params in Firefox 61, and Sync query records store the tag name in the `folderName` in their payloads, so it has enough information to fix these up.

Sync doesn't have enough information to fix up (2)—the receiving device doesn't know what folder ID the query wants—so it compromises by adding `excludeItems=1` to the query. The device that originally uploaded the query has to rewrite it from `folder=<folder ID> to `parent=<folder GUID>`, but, in the meantime, adding `excludeItems=1` will make sure the query doesn't return anything, to keep it from breaking the UI.

Comment (bendk) [reply components/places/src/bookmark_sync/tests/payload_evolution.rs:99]: Wow, that history is totally wild.  I copied your code and was able to get the remote_query test to fail properly and then was able to fix it.  I think payload evolution is working for all bookmark types other than livemarks, thanks for all the help!

File: components/places/src/bookmark_sync/incoming.rs [hunk 3]
```diff
@@ -38,31 +40,21 @@ impl<'a> IncomingApplicator<'a> {
         // This is a little odd - we need to go via serde_json::Value so we can
         // work out what we want to deserialize it as.
         let timestamp = record.envelope.modified;
-        let json_content = record.into_content::<serde_json::Value>();
+        let json_content = record.into_content::<BookmarkItemRecord>();
```
Comment (linabutler) [inline components/places/src/bookmark_sync/incoming.rs:43]: question: Will this still parse and flag records with invalid (wrong type or missing) properties, or will those now fall through to the `IncomingKind::Malformed` case? And, is that a case we should still be worried about?

If memory serves right, we switched to treating records as dynamically-checked `JsonValue` bags in #1661, which is also how Desktop behaves. We don't fail the entire sync anymore, but I'm wondering if we could end up in a situation where Desktop would fix up (replace / reupload) an invalid record, but Rust Places would treat that record as malformed and skip it.

Maybe keeping an eye on the `malformed-incoming-bookmark` probe is enough...

File: components/places/src/storage/bookmarks.rs [hunk 4]
```diff
@@ -907,13 +907,9 @@ fn get_raw_bookmarks_for_url(db: &PlacesDb, url: &Url) -> Result<Vec<RawBookmark
 }
 
 fn reset_in_tx(db: &PlacesDb, assoc: &EngineSyncAssociation) -> Result<()> {
-    // Remove all synced bookmarks and pending tombstones, and mark all
-    // local bookmarks as new.
+    // Mark all local bookmarks as new so that we re-upload them on the next sync
     db.execute_batch(&format!(
-        "DELETE FROM moz_bookmarks_synced;
-
-        DELETE FROM moz_bookmarks_deleted;
```
Comment (linabutler) [inline components/places/src/storage/bookmarks.rs:915]: question: How come we want to carry over unknown fields between resets? Wouldn't re-downloading all records on the first sync after a reset also download those unknown fields?

Comment (mhammond) [reply components/places/src/storage/bookmarks.rs:915]: One argument for keeping them is just that we might be the first device to show up after a reset, right?

But although I haven't looked much at this patch, it does seem a red-flag that a reset isn't killing the mirror.

Comment (mhammond) [reply components/places/src/storage/bookmarks.rs:915]: (I haven't put as much thought into resets as I should. Eg, I don't think autofill will keep them because it relies on what's stored in the mirror and deletes it on reset). It would seem odd for some engines to carry them and not others though.)

File: components/places/src/bookmark_sync/incoming.rs (hunk 3 above)
Comment (bendk) [reply components/places/src/bookmark_sync/incoming.rs:43]: I mentioned this in [the commit](https://github.com/mozilla/application-services/pull/5438/commits/8e1c752d08e182df52585c063b47f042414f1e68), but I forgot to call it out in the PR.  I think you're analysis is correct and it means we will skip over records with incorrect types / missing required fields.  I think that's okay, given the payload evolution discussions, but I wonder what other's think.

BTW, I also forgot to call out Mark's excellent new sync15 code that makes this all work.  It was a real joy doing that refactor.

File: components/places/src/storage/bookmarks.rs (hunk 4 above)
Comment (bendk) [reply components/places/src/storage/bookmarks.rs:915]: Maybe we need to define what a reset represents.  If reset means "forget the server data", then I think it should also mean that we forget the unknown fields.  If it means "reset your timestamps/counters so you do a full sync next time", then I think it's okay to keep the mirror data around.

File: components/places/src/bookmark_sync/incoming.rs (hunk 3 above)
Comment (mhammond) [reply components/places/src/bookmark_sync/incoming.rs:43]: > it means we will skip over records with incorrect types / missing required fields.

I'm a little torn here - bookmarks is somewhat special in that it has a fuzzy concept of requiring a kind of atomicity. Eg, [allow_failed](https://github.com/mozilla/application-services/blob/main/components/sync15/src/client/request.rs#L188) when posting exists only for bookmarks. The main concern is the possibility of creating an invalid tree - eg, if the POST creating a parent node failed, we don't want to allow the children. The same concern applies for incoming records - if the parent is malformed, we really can't skip it. So the decision to just give up was kinda explicit, but mainly in a "let's not think about this now" kind of way.

dogear *should* prevent us actually having a bad tree locally though - but if it really is just this device that considers the record malformed and dogear reparents items because of it, then that contagion ends up spreading to other devices too.

All that said though, one malformed record preventing bookmarks syncing *at all* on this device certainly doesn't seem an ideal outcome either. So there's no correct answer here, it's all a tradeoff of what we think might go wrong.

I'd love to hear what others think here.

File: components/places/src/storage/bookmarks.rs (hunk 4 above)
Comment (mhammond) [reply components/places/src/storage/bookmarks.rs:915]: It means "we have disconnected from the server", either explicitly due to user action, or because the server state changed (eg, password reset). In the former case, we may never reconnect, so keeping old server data around forever just smells wrong from a privacy POV - their DB still having bookmarks deleted years sounds quite bad. In a server-state change, then we are going to reconnect soon, but that differentiation doesn't currently exist. I guess we could consider looking into adding it.

File: components/places/src/bookmark_sync/incoming.rs (hunk 3 above)
Comment (mhammond) [reply components/places/src/bookmark_sync/incoming.rs:43]: *sigh* - I really should not try and comment without more context. I just deleted the above comment which totally missed the point - we do already skip malformed records as Lina noted. I'm actually not sure that was the right call for reasons noted by Lina and in my deleted comment. I wonder if we can change the "unpack" functions to take the json as `mut &` and them call [take](https://docs.rs/serde_json/latest/serde_json/value/enum.Value.html#method.take), then any non-null values left are unknown?

(Or we assert that the last 5 years *has* cleaned up things such that we can trust data better, however issues like https://mozilla.sentry.io/issues/3796658061 worry me - sentry hates me, but we see some reports with text `Invalid incoming T: invalid type: string "1676479234.51", expected i64` as well as null confusion in `Invalid incoming T: invalid type: null, expected a string`. We can be reasonably sure that the `i64` problem is *not* bookmarks, because `unpack_optional_i64` handles that case - so those kinds of records would regress here)

Comment (bendk) [reply components/places/src/bookmark_sync/incoming.rs:43]: Actually, I don't think that will work for this purpose since the functions were doing much more complicated things like trying to convert strings to ints and also marking when a record needs to be reuploaded.

**Comment by bendk** [comment]
Pushed a new version of this based on the conversation here.  There's two main changes:

- Based on the slack convo, I removed the commit to avoid deleting data from the mirror table and removed the test for round-tripping data after a reset.
- I went back to the older code that has workarounds for mistyped data.  I'd love to know what others think about this, I'm not totally sure about it myself.

File: components/places/src/bookmark_sync/incoming.rs [hunk 5]
```diff
@@ -521,83 +471,122 @@ impl<'a> IncomingApplicator<'a> {
...
+///
+/// This is extra important since bookmarks form a tree.  If a parent node is invalid, then we will
+/// have issues tring to merge its children.
+fn fixup_bookmark_json(json: &mut JsonValue) -> SyncedBookmarkValidity {
+    let mut validity = SyncedBookmarkValidity::Valid;
+    // the json value should always be on object, if not don't try to do any fixups.  The result will
+    // be that into_content_with_fixup() returns an IncomingContent with IncomingKind::Malformed.
+    if let JsonValue::Object(ref mut obj) = json {
+        obj.entry("parentid")
+            .and_modify(|v| fixup_optional_str(v, &mut validity));
```
Comment (bendk) [inline components/places/src/bookmark_sync/incoming.rs:485]: Before there was a an `unpack_optional_id` that worked like `unpack_optional_str`, but didn't modify the validity.  I don't see why we wouldn't want to re-upload when the `parentid` field is wrong, so I made them use the same codepath.

File: components/places/src/bookmark_sync/incoming.rs [hunk 6]
```diff
@@ -521,83 +471,122 @@ impl<'a> IncomingApplicator<'a> {
...
+                        log::trace!("Incoming bookmark has duplicate tag: {:?}", tag);
+                        set_reupload(validity);
+                    }
+                } else {
+                    log::trace!("Incoming bookmark has unexpected tag: {:?}", v);
+                    set_reupload(validity);
+                }
+            }
+            if valid_tags.len() != tags.len() {
+                *val = JsonValue::Array(valid_tags.into_iter().map(JsonValue::from).collect());
```
Comment (linabutler) [inline components/places/src/bookmark_sync/incoming.rs:583]: nitpick: Let's do this unconditionally; I think `valid_tags` might still have normalized tags even when `valid_tags.len() == tags.len()`?

File: components/places/src/bookmark_sync/incoming.rs [hunk 7]
```diff
@@ -521,83 +471,122 @@ impl<'a> IncomingApplicator<'a> {
...
-    match &data[key] {
-        JsonValue::String(ref s) => {
-            // Like Desktop, we don't reupload if a keyword has leading or
-            // trailing whitespace, or isn't lowercase.
-            let k = s.trim();
-            if k.is_empty() {
-                None
-            } else {
-                Some(k.to_lowercase())
+fn fixup_optional_tags(val: &mut JsonValue, validity: &mut SyncedBookmarkValidity) {
```
Comment (linabutler) [inline components/places/src/bookmark_sync/incoming.rs:555]: Thanks for untangling this logic from `store_incoming_bookmark`! 🏆

File: components/places/src/bookmark_sync/tests/payload_evolution.rs [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        )],
+        outgoing_unknown_fields: vec![
+            (guid('a'), vec![("device_type", "mini-phone")]),
+            (guid('b'), vec![("device_type", "desktop")]),
+        ],
+    }
+    .run()
+}
+
+struct RoundtripTest {
```
Comment (linabutler) [inline components/places/src/bookmark_sync/tests/payload_evolution.rs:103]: ❤️ table-driven tests!

**Review by linabutler — APPROVED** [review]
LGTM!

File: components/places/src/bookmark_sync/tests/payload_evolution.rs (hunk 8 above)
Comment (bendk) [reply components/places/src/bookmark_sync/tests/payload_evolution.rs:103]: Yeah, they're pretty great.  I saw a few in the app-services code when I first came on and copied the style.  The more I write the more I like them.

File: components/places/sql/create_shared_schema.sql [hunk 9]
```diff
@@ -196,7 +196,9 @@ CREATE TABLE IF NOT EXISTS moz_bookmarks_synced(
     loadInSidebar BOOLEAN,
     smartBookmarkName TEXT,
     feedURL TEXT,
-    siteURL TEXT
+    siteURL TEXT,
+    -- All unknown fields from the server record, encoded as a JSON object.
+    unknownFields TEXT
```
Comment (mhammond) [inline components/places/sql/create_shared_schema.sql:201]: This needs a db migration, right?

File: components/places/src/bookmark_sync/incoming.rs [hunk 10]
```diff
@@ -130,12 +96,13 @@ impl<'a> IncomingApplicator<'a> {
                 None
             }
         };
+        let unknown_fields = serde_json::to_string(&b.unknown_fields)?;
```
Comment (mhammond) [inline components/places/src/bookmark_sync/incoming.rs:99]: How do you feel about writing a null if the map is empty? (And ditto for other items below)

**Review by mhammond — COMMENTED** [review]
This is a great refactor, thanks. I think storing `NULL` seems optimal, but not critical, however the DB migration is. Probably worth running the `places_utils` example too to ensure the upgrade works and that syncing still works after.
